### PR TITLE
Mid: attrd: Support of the dampen change by attrd.

### PR DESCRIPTION
**Pull request by HideoYamauchi**
Hi All,

I added the support of the change of dampen by attrd and attrd_updater.
I added a function to change the value only for dampen and a function to change the combination of dampen and value.

The command-line of attrd_updater is as follows.

1)Change the combination of dampen and value.
  attrd_updater -E value -n name -d dampen

2)Change the value only for dampen.
  attrd_updater -F -n name -d dampen

It is necessary to discuss the next matter.
 * Option of the attrd_updater command.
  * The current patch uses E and F by tentativeness.
 * Help of attrd_updater.
  * Because I am weak in English, I do not understand the expression of the right help.
 * Lack of the processing may be to the patch.

I make this patch from former next pullrequest.
 * https://github.com/ClusterLabs/pacemaker/pull/906

Best Regards,
Hideo Yamauchi.

### CODE REVIEW

File: attrd/commands.c [hunk 1]
```diff
@@ -669,14 +669,41 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
...
     const char *value = crm_element_value(xml, F_ATTRD_VALUE);
+    const char *dvalue = crm_element_value(xml, F_ATTRD_DAMPEN);
 
     attribute_t *a = g_hash_table_lookup(attributes, attr);
 
     if(a == NULL) {
-        a = create_attribute(xml);
+        if (safe_str_eq(op, ATTRD_OP_UPDATE) || safe_str_eq(op, ATTRD_OP_UPDATE_VALUE_AND_DAMPEN)) {
+            a = create_attribute(xml);
+        }
```
Comment (kgaillot) [inline attrd/commands.c]: Hi Hideo,

Thanks for contributing the new patch. It is much appreciated.

Instead of the above "if", I think we should print an error and return, if "dampen only" was requested. That way, the rest of the code can continue to assume that "a" is not NULL.

File: attrd/commands.c [hunk 2]
```diff
@@ -669,14 +669,41 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
...
+                            crm_info("Update attribute %s with delay %dms (%s)", a->id, dampen, dvalue);
+                        } else {
+                            a->timer = NULL;
+                            crm_info("Update attribute %s with not delay", a->id);
+                        }
+                        //if dampen is changed, attrd writes in a current value immediately.
+                        write_or_elect_attribute(a);
+                    }
+                } 
+            }
```
Comment (kgaillot) [inline attrd/commands.c]: Let's print an error if dampen does not exist or is not >=0.

File: attrd/commands.c [hunk 3]
```diff
@@ -669,14 +669,41 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
...
+        }
```
Comment (kgaillot) [inline attrd/commands.c]: If "dampen only" was requested, we can return after changing the dampen. That would avoid having to change the later code.

File: tools/attrd_updater.c [hunk 4]
```diff
@@ -45,6 +45,8 @@ static struct crm_option long_options[] = {
     {"-spacer-",1, 0, '-', "\nCommands:"},
     {"update",  1, 0, 'U', "Update the attribute's value in attrd.  If this causes the value to change, it will also be updated in the cluster configuration"},
 #ifdef HAVE_ATOMIC_ATTRD
+    {"update-all",   1, 0, 'E', "Update the attribute's value, and the time to wait (dampening) in attrd.  If this causes the value to change, it will also be updated in the cluster configuration"},
```
Comment (kgaillot) [inline tools/attrd_updater.c]: I'm thinking something like this:
{"update-both",   1, 0, 'B', "Update the attribute's value and time to wait (dampening) in attrd. If this causes the value or dampening to change, the attribute will also be written to the cluster configuration, so be aware that repeatedly changing the dampening reduces its effectiveness."}

File: tools/attrd_updater.c [hunk 5]
```diff
@@ -45,6 +45,8 @@ static struct crm_option long_options[] = {
...
+    {"update-dampen",   0, 0, 'F', "Update the time to wait (dampening) in attrd. If this causes the value to change, it will also be updated in the cluster configuration"},
```
Comment (kgaillot) [inline tools/attrd_updater.c]: How about:
{"update-delay",   1, 0, 'Y', "Update the attribute's dampening in attrd (requires -d/--delay). If this causes the dampening to change, the attribute will also be written to the cluster configuration, so be aware that repeatedly changing the dampening reduces its effectiveness."},

File: attrd/commands.c (hunk 1 above)
Comment (HideoYamauchi) [inline attrd/commands.c]: Hi Ken,

Okay.
I add the following correction.
`
        if (safe_str_eq(op, ATTRD_OP_UPDATE_ONLY_DAMPEN)) {
            crm_notice("Update error (attribute %s not found)", attr);
            return;
        }
`
Is my understanding wrong?

File: attrd/commands.c (hunk 3 above)
Comment (HideoYamauchi) [inline attrd/commands.c]: Some these corrections were troubled with me.
I add the following correction.

``
                        //The change only for dampen is late to the change of the next attribute or the next timer
                        if (safe_str_eq(op, ATTRD_OP_UPDATE_ONLY_DAMPEN)) {
                            return;
                        }
                        //if dampen is changed, attrd writes in a current value immediately.
                        write_or_elect_attribute(a);
``

File: tools/attrd_updater.c (hunk 4 above)
Comment (HideoYamauchi) [inline tools/attrd_updater.c]: Splendid.

I accept your suggestion.
I intend to change ATTRD_OP_UPDATE_VALUE_AND_DAMPEN to　ATTRD_OP_UPDATE_BOTH.

File: tools/attrd_updater.c (hunk 5 above)
Comment (HideoYamauchi) [inline tools/attrd_updater.c]: Okay!

I accept your suggestion.
I intend to change ATTRD_OP_UPDATE_DAMPEN to　ATTRD_OP_UPDATE_DELAY.

**Comment by HideoYamauchi** [comment]
Hi Ken,

Your suggestion waits for the suggestion from other members during a little and reflects it.
I am going to make the patch which reflected your suggestion in the first half of the next week.

Many thanks!
Hideo Yamauchi.

File: attrd/commands.c (hunk 2 above)
Comment (HideoYamauchi) [inline attrd/commands.c]: Okay!
I add the following correction.

``
            if (dvalue) {
                dampen = crm_get_msec(dvalue); 
                if (dampen >= 0) {
                    if (a->timeout_ms != dampen) {
(snip)
                } else {
			crm_notice("Update error (A positive number is necessary for delay parameter. %dms (%s))", dampen, dvalue);
			return;
		}
(snip)
            } else {
		crm_notice("Update error (delay parameter is necessary for the update of the attribute %s)", attr);
		return;
	    }

``

Is my understanding wrong?

File: attrd/commands.c (hunk 1 above)
Comment (kgaillot) [inline attrd/commands.c]: That is correct. I would use crm_warn() instead of crm_notice().

File: attrd/commands.c (hunk 2 above)
Comment (kgaillot) [inline attrd/commands.c]: Your fix looks good. Again, I would use crm_warn() instead of crm_notice() here.

Good point about create_attribute(). Maybe: ... else if (dampen < 0) { crm_warn("Ignoring invalid delay %s for attribute %s" ...

File: attrd/commands.c (hunk 3 above)
Comment (kgaillot) [inline attrd/commands.c]: I think with dampen only, we do not need any further processing:

if (safe_str_eq(op, ATTRD_OP_UPDATE_ONLY_DAMPEN)) {
    return;
}
write_or_elect_attribute(a);

Then, the changes to the next part of this function would be unnecessary.

**Comment by kgaillot** [comment]
Hi Hideo,

That sounds good. I will wait for the new patch.

File: attrd/commands.c (hunk 1 above)
Comment (HideoYamauchi) [inline attrd/commands.c]: Okay!

File: attrd/commands.c (hunk 2 above)
Comment (HideoYamauchi) [inline attrd/commands.c]: Okay!

I send a different patch to the correction of create_attribute().

File: attrd/commands.c (hunk 3 above)
Comment (HideoYamauchi) [inline attrd/commands.c]: When attrd throws processing up in the case of ATTRD_OP_UPDATE_ONLY_DAMPEN, in the case of next, an attribute is not updated.
It is not updated at all until an attribute is changed again.

1. I send the attribute together first in damepn=60.(attrd_updater -n test -v XXX -d 60)
1. While it does not pass for 60 seconds, I change dampen.(attrd_updater -Y -d 20)
1. As a result, because I changed dampen without waiting for the progress of first 60 seconds, the attribute continues being maintained until there is the change of the next attribute.(cib will not have test attribute until the change of the next value.)

What do you think?

**Comment by HideoYamauchi** [comment]
Hi Ken,

About processing only for dampen, please tell me the opinion.

```
(snip)
1. I send the attribute together first in damepn=60.(attrd_updater -n test -v XXX -d 60)
2. While it does not pass for 60 seconds, I change dampen.(attrd_updater -Y -d 20)
3. As a result, because I changed dampen without waiting for the progress of first 60 seconds, the attribute continues being maintained until there is the change of the next attribute.(cib will not have test attribute until the change of the next value.)
(snip)
```
Best Regards,
Hideo Yamauchi.

File: attrd/commands.c (hunk 3 above)
Comment (kgaillot) [inline attrd/commands.c]: I see -- that makes sense. What do you think of this:

write_or_elect_attribute(a);
if (safe_str_eq(op, ATTRD_OP_UPDATE_ONLY_DAMPEN)) {
return;
}

Then, we don't need the changes to the rest of this function.

Comment (HideoYamauchi) [inline attrd/commands.c]: Okay!
I think that it is right.

**Comment by HideoYamauchi** [comment]
Hi Ken,

Thank you for comment.
The problem became clear.

I am going to send the patch which reflected your opinion early next week.

Many thanks!
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
The new patch was ready.
This does Close.
